src/NuGet.Clients/VsExtension/VisualStudioAccountProvider.cs

```diff
@@ -0,0 +1,390 @@
+                    {
+                        accountsWithAccess.Add(new AccountWithCreds(account, cred));
+                    }
+                }
+
+                if (accountsWithAccess.Count == 1)
+                {
+                    ret = accountsWithAccess[0].Creds;
+                }
+                else
+                {
+                    // we couldn't finde a unique account with access to the endpoint so we are going to have
+                    // to ask the user...
+                    ret = await PromptUserForAccount(uriTenantId, provider, nonInteractive, cancellationToken)
+                        .ConfigureAwait(false);
+                }
+
+            }
+            else // count == 0 so we should prompt the user
+            {
+                ret = await PromptUserForAccount(uriTenantId, provider, nonInteractive, cancellationToken)
+                    .ConfigureAwait(false);
+            }
+
+            if (ret == null)
+            {
+                // No credentials found but we know that this is a VSO endpoint so we want to throw an
+                // exception to prevent the other providers from being called
+                throw new Exception("No valid credentials found for VSO account");
+            }
+
+            return ret;
+        }
+
+        // Logic shows UI and interacts with all mocked methods.  Mocking this as well.
+        internal virtual async Task<ICredentials> PromptUserForAccount(
+        	string tenentId, VSAccountProvider provider, bool nonInteractive, CancellationToken cancellationToken)
+        {
+            ICredentials ret = null;
+            if (nonInteractive)
+            {
+                //  If we are not supposed to interact with the user then we can't prompt for account so we
+                // need to fail.
+                return null;
+            }
+            Account account = null;
+
+            ThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                var parent = IntPtr.Zero;
+                if (_dte != null)
+                {
+                    parent = new IntPtr(_dte.MainWindow.HWnd);
+                }
+
+                account = await provider.CreateAccountWithUIAsync(parent, cancellationToken)
+                    .ConfigureAwait(false);
+            });
+
+            var tenant = FindTenantInAccount(account, tenentId, provider);
+            if (tenant != null)
+            {
+                ret = await GetTokenFromAccount(
+                    new AccountAndTenant(account, tenant),
+                    provider,
+                    false,
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            return ret;
+        }
+
+        // Logic goes between UI and web.  made internal to be mocked for unit tests
+        internal virtual async Task<ICredentials> GetTokenFromAccount(AccountAndTenant account,
```

Comment:
What pattern do you recommend we use for mocking if you don't want us using InternalsVisibleTo? We could certainly express this functionality in an interface and have the it passed into VisualStudioAccountProvider at construction, but it seems like overkill.